triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.oauth.middleware
+
+import akka.Done
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.Http.ServerBinding
+import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport._
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.headers.HttpCookie
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.unmarshalling.{Unmarshal, Unmarshaller}
+import com.daml.oauth.server.{Request => OAuthRequest, Response => OAuthResponse}
+import com.typesafe.scalalogging.StrictLogging
+import java.util.Base64
+import java.util.UUID
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.postfixOps
+import scala.util.Try
+import spray.json._
+
+// This is an implementation of the trigger service authentication middleware
+// for OAuth2 as specified in `/triggers/service/authentication.md`
+object Server extends StrictLogging {
+  import com.daml.oauth.server.JsonProtocol._
+
+  private def toRedirectUri(uri: Uri) =
+    Uri()
+      .withScheme(uri.scheme)
+      .withAuthority(uri.authority)
+      .withPath(Uri.Path./("cb"))
+
+  def start(
+      config: Config)(implicit system: ActorSystem, ec: ExecutionContext): Future[ServerBinding] = {
+    implicit val unmarshal: Unmarshaller[String, Uri] = Unmarshaller.strict(Uri(_))
+    // TODO[AH] Make sure this is bounded in size - or avoid state altogether.
+    val requests = TrieMap[UUID, Uri]()
+    val route = concat(
+      path("auth") {
+        get {
+          complete((StatusCodes.NotImplemented, "The /auth endpoint is not implemented yet"))
+        }
+      },
+      path("login") {
+        get {
+          parameters(('redirect_uri.as[Uri], 'claims))
+            .as[Request.Login](Request.Login) {
+              login =>
+                extractRequest {
+                  request =>
+                    val requestId = UUID.randomUUID
+                    requests += (requestId -> login.redirectUri)
+                    val authorize = OAuthRequest.Authorize(
+                      responseType = "code",
+                      clientId = config.clientId,
+                      redirectUri = toRedirectUri(request.uri),
+                      scope = Some(login.claims),
```

Comment:
We probably need to make the scopes configurable to some degree. Some oauth providers (e.g. auth0) also allow you to use custom parameters instead of scopes to accomplish the same thing. Not quite sure what the best UX for that is and we don’t have to solve this in this PR but something to keep in mind for the future.

Reply:
That's good to know, the implementation of that is probably best guided by a concrete use-case that requires that sort of thing. The mapping from claims to scope generally still needs work.